pkg/http/universalapi.go

```diff
@@ -51,10 +59,109 @@ type UniversalFastHTTPHandlerOpts[T proto.Message, U proto.Message] struct {
 	ProtoResponseEmitUnpopulated bool
 }
 
+// UniversalFastHTTPHandler wraps a Universal API method into a HTTP handler.
+func UniversalHTTPHandler[T proto.Message, U proto.Message](
+	handler func(ctx context.Context, in T) (U, error),
+	opts UniversalHTTPHandlerOpts[T, U],
+) func(http.ResponseWriter, *http.Request) {
+	var zero T
+	rt := reflect.ValueOf(zero).Type().Elem()
+	pjsonDec := protojson.UnmarshalOptions{
+		DiscardUnknown: true,
+		AllowPartial:   true,
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		var err error
+
+		// Need to use some reflection magic to allocate a value for the pointer of the generic type T
+		in := reflect.New(rt).Interface().(T)
+
+		// Parse the body as JSON
+		if !opts.SkipInputBody && r.Body != nil {
+			var body []byte
+			// Read the request body and decode it as JSON using protojson
+			body, err = io.ReadAll(r.Body)
+			if err != nil {
+				msg := messages.ErrBodyRead.WithFormat(err)
+				respondWithError(w, msg)
+				log.Debug(msg)
+				return
+			}
+
+			if len(body) > 0 {
+				err = pjsonDec.Unmarshal(body, in)
+				if err != nil {
+					msg := messages.ErrMalformedRequest.WithFormat(err)
+					respondWithError(w, msg)
+					log.Debug(msg)
+					return
+				}
+			}
+		}
```

Comment:
Is there no way to marshal the body reader directly? Pain to read the entire body first.

Reply:
:+1: sadge